R/relevel.R

```diff
@@ -78,3 +81,128 @@ wrap_expr_in_expr <- function(pd) {
+#' `EQ_ASSIGN` and the tokens to its left and right are located too high up in
+#' the hierarchy of the nested parse data. Hence, this function wraps the
+#' sub-expression into an expression, similar to [wrap_expr_in_curly()].
+#' Since `wrap_expr_in_curly()` is called from within a visitor
+#' (and `relocate_eq_assign()` not), we need to
+#' wrap the the implementation [relocate_eq_assign_nest()] that operates on
+#' *nests* into a visitor call.
+#' @param pd A parse table.
+#' @examples
+#' styler:::get_parse_data("a <- b <- 3")
+#' styler:::get_parse_data("a  = b = 3")
+#' styler:::get_parse_data(
+#'   "x = 5
+#'   if(x >= 5)
+#'   y = TRUE else
+#'   y = FALSE",
+#' )
+#' styler:::get_parse_data(
+#'   "x <- 5
+#'   if(x >= 5)
+#'   y <- TRUE else
+#'   y <- FALSE",
+#' )
+relocate_eq_assign <- function(pd) {
+  pd %>%
+    post_visit(c(relocate_eq_assign_nest))
+}
+
+
+#' Relocate all assignment expressions that contain `EQ_ASSIGN` within a *nest*
+#'
+#' Implements the relocation of an `EQ_ASSIGN` and associated tokens
+#' within a *nest* (nested parse table at one level of nesting).
+#' Note that one assignment expression (such as "a = b = c") can include
+#' multiple assignment operators, an assignment involves just one assignment
+#' oparator.
+#' For the relocation of assignment expressions that contain `EQ_ASSIGN` within
+#' a *nest*, we need to first find the expressions that contain `=` and then
+#' split the *nest* into parse tables each containing one such assignment
+#' expression and then relocate each of them separately.
+#' We can't do all of them together because:
+#'
+#'  * An assignment can contain more than just three tokens, e.g. (a <- b <- c).
+#'  * Two assignemnts can be in the same nest although they don't belong to the
+#'    same assignment (if-else statement).
+#'
+#' Please refer to the section 'Examples' in [relocate_eq_assign()] for details.
+#' @param pd A parse table.
+#' @importFrom rlang seq2
+#' @importFrom purrr map_dfr
+relocate_eq_assign_nest <- function(pd) {
+  is_eq_assign <- which(pd$token == "EQ_ASSIGN")
+  if (length(is_eq_assign) > 0) {
+    block_id <- find_block_id(pd)
+    blocks <- split(pd, block_id)
+    pd <- map_dfr(blocks, relocate_eq_assign_one)
+  }
+  pd
+}
+
+
+#' Find the block to which a token belongs
+#'
+#' Two assignment tokens `EQ_ASSIGN` belong to the same block if they are not
+#' separated by more than one token. Token between `EQ_ASSIGN` tokens belong
+#' to the `EQ_ASSIGN` token occuring before them, except the token right before
+#' `EQ_ASSING` already belongs to the `EQ_ASSING` after it.
+#' @param pd A parse table.
+find_block_id <- function(pd) {
+  is_eq_assign <- which(pd$token == "EQ_ASSIGN")
+  eq_belongs_to_block <-
+    cumsum((is_eq_assign - lag(is_eq_assign, default = is_eq_assign[1])) > 2)
+  empty_seq <- rep(0, nrow(pd))
+  empty_seq[lead(pd$token == "EQ_ASSIGN", default = FALSE)] <- eq_belongs_to_block
+  block_id <- cumsum(empty_seq)
```

Comment:
Is this the value this function returns?

Reply:
Yup. Lets add 
```r
block_id
```
So it is explicit.